test/OtokenFactory.test.ts

```diff
@@ -0,0 +1,221 @@
+const MockOtoken = artifacts.require('MockOtoken.sol')
+const ZERO_ADDR = '0x0000000000000000000000000000000000000000'
+
+contract('OTokenFactory', accounts => {
+  let otoken: MockOtokenInstance
+  let addressBook: MockAddressBookInstance
+  let otokenFactory: OtokenFactoryInstance
+
+  // Paramter used for otoken init(). (Use random addresses as usdc and eth)
+  let usdc: MockERC20Instance
+  let shitcoin: MockERC20Instance
+  const ethAddress = ZERO_ADDR
+  const strikePrice = new BigNumber(200).times(new BigNumber(10).exponentiatedBy(18))
+  const isPut = true
+  let expiry: number
+
+  before('Deploy otoken logic and Factory contract', async () => {
+    usdc = await MockERC20.new('USDC', 'USDC')
+    shitcoin = await MockERC20.new('Shit coin', 'STC')
+
+    const logic = await MockOtoken.new()
+
+    // Deploy and whitelist ETH:USDC product
+    const mockWhitelist: MockWhitelistModuleInstance = await MockWhitelist.new()
+    await mockWhitelist.whitelistProduct(ethAddress, usdc.address, usdc.address)
+    // Deploy addressbook
+    addressBook = await MockAddressBook.new()
+    await addressBook.setOtokenImpl(logic.address)
+    await addressBook.setWhitelist(mockWhitelist.address)
+
+    otokenFactory = await OTokenFactory.new(addressBook.address)
+    expiry = (await time.latest()).toNumber() + time.duration.days(30).toNumber()
+  })
+
+  describe('Get otoken address', () => {
+    it('Should have no otoken records at the begining', async () => {
+      const otokens = await otokenFactory.getOtokens()
+      assert.equal(otokens.length, 0, 'Should have no otoken records')
+    })
+
+    it('Should return address(0) if token is not deployed', async () => {
+      const existAddress = await otokenFactory.getOtoken(
+        ethAddress,
+        usdc.address,
+        usdc.address,
+        strikePrice,
+        expiry,
+        isPut,
+      )
+      assert.equal(existAddress, ZERO_ADDR, 'getOtoken check failed on undeployed tokens.')
+    })
+
+    it('should get deterministic address with new otoken paramters', async () => {
+      const targetAddress = await otokenFactory.getTargetOtokenAddress(
+        ethAddress,
+        usdc.address,
+        usdc.address,
+        strikePrice,
+        expiry,
+        isPut,
+      )
+      assert.notEqual(targetAddress, ZERO_ADDR, 'getTargetOtokenAddress should never give 0 address.')
+    })
+
+    it('should get different target address with different otoken paramters', async () => {
+      const targetAddress1 = await otokenFactory.getTargetOtokenAddress(
+        ethAddress,
+        usdc.address,
+        usdc.address,
+        strikePrice,
+        expiry,
+        isPut,
+      )
+      const targetAddress2 = await otokenFactory.getTargetOtokenAddress(
+        ZERO_ADDR,
+        ZERO_ADDR,
+        ZERO_ADDR,
+        strikePrice,
+        expiry,
+        isPut,
+      )
+      assert.notEqual(targetAddress1, targetAddress2)
+    })
+  })
+
+  describe('Create new otoken', () => {
```

Comment:
can you test that anyone can create these? Ensure two different addresses can create a valid oToken?